data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java

```diff
@@ -60,58 +63,78 @@ public class DspCatalogApiController {
         this.mapper = mapper;
         this.identityService = identityService;
         this.transformerRegistry = transformerRegistry;
         this.dspCallbackAddress = dspCallbackAddress;
         this.service = service;
         this.jsonLdService = jsonLdService;
+        this.statusCodeMapper = statusCodeMapper;
     }
 
     @POST
     @Path(CATALOG_REQUEST)
-    public Map<String, Object> getCatalog(JsonObject jsonObject, @HeaderParam(AUTHORIZATION) String token) {
+    public Response getCatalog(JsonObject jsonObject, @HeaderParam(AUTHORIZATION) String token) {
         monitor.debug(() -> "DSP: Incoming catalog request.");
 
-        var tokenRepresentation = TokenRepresentation.Builder.newInstance()
-                .token(token)
-                .build();
-
-        var claimToken = identityService.verifyJwtToken(tokenRepresentation, dspCallbackAddress)
-                .orElseThrow(failure -> new AuthenticationFailedException());
-
-        var expanded = jsonLdService.expand(jsonObject);
-        if (expanded.failed()) {
-            throw new InvalidRequestException(expanded.getFailureDetail());
+        try {
+            var tokenRepresentation = TokenRepresentation.Builder.newInstance()
+                    .token(token)
+                    .build();
+
+            var claimToken = identityService.verifyJwtToken(tokenRepresentation, dspCallbackAddress)
+                    .orElseThrow(failure -> new AuthenticationFailedException());
+
+            var expanded = jsonLdService.expand(jsonObject);
+            if (expanded.failed()) {
+                throw new InvalidRequestException(expanded.getFailureDetail());
+            }
+            var expandedJson = expanded.getContent();
+            if (!isOfExpectedType(expandedJson, DSPACE_CATALOG_REQUEST_TYPE)) {
+                throw new InvalidRequestException(format("Request body was not of expected type: %s", DSPACE_CATALOG_REQUEST_TYPE));
+            }
+
+            var message = transformerRegistry.transform(expandedJson, CatalogRequestMessage.class)
+                    .orElseThrow(failure -> new InvalidRequestException(format("Request body was malformed: %s", failure.getFailureDetail())));
+
+            // set protocol
+            message.setProtocol(DATASPACE_PROTOCOL_HTTP);
+
+            var catalog = service.getCatalog(message, claimToken)
+                    .orElseThrow(exceptionMapper(Catalog.class));
+
+            var catalogJson = transformerRegistry.transform(catalog, JsonObject.class)
+                    .orElseThrow(failure -> new EdcException(format("Failed to build response: %s", failure.getFailureDetail())));
+
+            var compacted = jsonLdService.compact(catalogJson);
+            if (compacted.failed()) {
+                throw new InvalidRequestException(compacted.getFailureDetail());
+            }
+            //noinspection unchecked
+            var result = mapper.convertValue(compacted.getContent(), Map.class);
```

Comment:
I always wondered... do we need this conversion? jersey should be able to serialize `JsonObject` correctly without converting it to a map before. Maybe something for another issue/PR